packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -557,162 +558,163 @@ impl<'m> RunnerImpl<'m> {
         // `data_node_from_js` isn't recursive -- doesn't convert children.
         let data: mv8::DataFfi = mv8.data_node_from_js(data);
 
-        let n_children = child_data.len();
-        let child_data: Vec<ArrayData> = match dt.clone() {
-            DataType::List(field) => {
-                let child: mv8::Value<'_> = child_data.get(0)?;
-                Ok(vec![self.array_data_from_js(
+        // This target length is used because the JS repr does not mirror buffer building as Rust
+        // Arrow and pyarrow.
+        let target_len = len.unwrap_or(data.len);
+
+        // TODO: We currently copy the buffers because the JavaScript representation of arrays does
+        //   not match the Rust implementation. Try to reduce copies where possible.
+        let mut builder = ArrayData::builder(field.data_type().clone());
+
+        match field.data_type() {
+            DataType::Boolean => {
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `u8`, and
+                //   `target_len` is carefully chosen
+                let values = unsafe { slice::from_raw_parts(data.buffer_ptrs[0], target_len) };
+
+                // Booleans are packed in arrow, `values` are already packed, so use them directly
+                let mut boolean_builder = BooleanBufferBuilder::new(target_len);
+                boolean_builder.append_packed_range(0..target_len, values);
+
+                builder = builder.add_buffer(boolean_builder.finish());
+            }
+            DataType::UInt16 => {
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `u16`, and
+                //   `target_len` is carefully chosen
+                let values =
+                    unsafe { slice::from_raw_parts(data.buffer_ptrs[0] as *const u16, target_len) };
+
+                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
+            }
+            DataType::UInt32 => {
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `u32`, and
+                //   `target_len` is carefully chosen
+                let values =
+                    unsafe { slice::from_raw_parts(data.buffer_ptrs[0] as *const u32, target_len) };
+
+                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
+            }
+            DataType::Float64 => {
+                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `f64`, and
+                //   `target_len` is carefully chosen
+                let values =
+                    unsafe { slice::from_raw_parts(data.buffer_ptrs[0] as *const f64, target_len) };
+
+                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
+            }
+            DataType::Utf8 => {
+                // Utf8 is stored in two buffers:
+                //   [0]: The offset buffer (i32)
+                //   [1]: The value buffer (u8)
+
+                // TODO: Use `data.len` or target_len?
```

Comment:
Do we want to get rid of this TODO? You had `data.len` in some places and it caused errors didn't it?